src/playground/UserConnection.hack

```diff
@@ -0,0 +1,45 @@
+use namespace Slack\GraphQL;
+use namespace HH\Lib\{Math, Str};
+
+final class UserConnection extends GraphQL\Pagination\Connection<User> {
+    protected async function paginate(
+        GraphQL\Pagination\PaginationArgs $args,
+    ): Awaitable<vec<GraphQL\Pagination\Edge<User>>> {
+        $after = $args['after'] ?? null;
+        $start_id = 0;
+        if ($after) {
+            $start_id = Str\to_int($after) as nonnull + 1; // Add one to skip the `after` cursor.
+        }
+
+        $before = $args['before'] ?? null;
+        $end_id = 5;
+        if ($before) {
+            $end_id = Str\to_int($before) as nonnull;
+        }
+
+        $first = $args['first'] ?? null;
+        if ($first) {
+            $end_id = Math\minva($end_id, $start_id + $first);
+        }
+
+        $last = $args['last'] ?? null;
+        if ($last) {
+            $start_id = Math\maxva($start_id, $end_id - $last);
+        }
+
+        $edges = vec[];
+        for ($i = $start_id; $i < $end_id; $i++) {
+            $edges[] = new GraphQL\Pagination\Edge(
```

Comment:
This feels ok to me! It would be nice to just have this return `Awaitable<vec<User>>` but I think the ability to paginate a simple vec will be useful.

Reply:
yeah. let's keep thinking about this. It feels OK to me too, but would love to find a way to support both.

Reply:
We need the implementor to provide a node and and a cursor for each item (+any extra fields if we support them in the future) and this seems like a pretty natural way to do that (it might be a bit nicer if we could take care of automatically encoding the cursor though -- not just for convenience but mainly to prevent bugs from people forgetting that).

But if we see people writing too much boilerplate around this, a possible option would be:

1. subclass returns a list of generic type TItem instead of a list of Edge
2. subclass implements a method that transforms TItem -> Edge (presumably there would be some common ones already implemented, so most developers only need to do 1 -- otherwise this wouldn't save much)